src/prefect/client/base.py

```diff
@@ -154,27 +155,69 @@ class PrefectHttpxClient(httpx.AsyncClient):
 
     RETRY_MAX = 5
 
-    async def send(self, *args, **kwargs) -> Response:
-        retry_count = 0
-        response = PrefectResponse.from_httpx_response(
-            await super().send(*args, **kwargs)
-        )
+    async def _send_with_retry(
+        self,
+        request: Callable,
+        retry_codes: Set[int] = {},
+        retry_exceptions: Tuple[Exception, ...] = (),
+    ):
+        """
+        Send a request and retry it if it fails.
+
+        Sends the provided request and retries it up to self.RETRY_MAX times if
+        the request either raises an exception listed in `retry_exceptions` or receives
+        a response with a status code listed in `retry_codes`.
+
+        Retries will be delayed based on either the retry header (preferred) or
+        exponential backoff if a retry header is not provided.
+        """
+        try_count = 0
+        response = None
+        retry = True
+
         while (
-            response.status_code
-            in {status.HTTP_429_TOO_MANY_REQUESTS, status.HTTP_503_SERVICE_UNAVAILABLE}
-            and retry_count < self.RETRY_MAX
+            not response
+            or (response.status_code in retry_codes)
```

Comment:
I am not certain but could this be:
```
while (not response or retry):
```

It looks like retry is being set on the same statements that the while loop is conditional on? I might be missing a case where the reverse doesn't make sense. Or maybe the `retry` variable is unneeded? I can't quite put my finger on it 😅

Reply:
Is the new arrangement more readable?

Reply:
I'm so sorry, I think my original comment was a little vague and unhelpful! I think a combination of the conditionals underneath the `try/except/else/finally` structure makes it harder to follow the code paths. Maybe something like this could work? (Just to qualify a bit I don't know if this works and the solution as written is great, just seeing if it's possible to consolidate a bit)

```
        try_count = 0
        response = None

        while True:
            try_count += 1
            retry_seconds = None
            retry = try_count <= self.RETRY_MAX

            try:
                response = await request()
            except retry_exceptions:
                if not retry:
                    raise
                retry_seconds = 2**try_count
            else:
                if response.status_code in retry_codes:
                    retry_seconds = float(response.headers.get("Retry-After", 2**try_count))

            if retry and retry_seconds is not None:
                await anyio.sleep(retry_seconds)
            else:
                break

        return response
```

Reply:
@jakekaplan want to give it a quick look again?